common/src/main/java/com/gregtechceu/gtceu/api/recipe/ShapedEnergyTransferRecipe.java

```diff
@@ -0,0 +1,125 @@
+package com.gregtechceu.gtceu.api.recipe;
+
+import com.google.gson.JsonObject;
+import com.gregtechceu.gtceu.api.capability.GTCapabilityHelper;
+import com.gregtechceu.gtceu.api.capability.IElectricItem;
+import com.gregtechceu.gtceu.core.mixins.ShapedRecipeInvoker;
+import lombok.Getter;
+import net.minecraft.FieldsAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+import net.minecraft.core.NonNullList;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.util.GsonHelper;
+import net.minecraft.world.inventory.CraftingContainer;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.minecraft.world.item.crafting.RecipeSerializer;
+import net.minecraft.world.item.crafting.ShapedRecipe;
+import net.minecraft.world.level.Level;
+
+import java.util.Map;
+
+/**
+ * @author Irgendwer01
+ * @date 2023/11/4
+ * @implNote ShapedEnergyTransferRecipe
+ */
+@MethodsReturnNonnullByDefault
+@FieldsAreNonnullByDefault
+public class ShapedEnergyTransferRecipe extends ShapedRecipe {
+    public static final RecipeSerializer<ShapedEnergyTransferRecipe> SERIALIZER = new Serializer();
+
+    private CraftingContainer craftingContainer;
+
+    @Getter
+    private final ItemStack chargeStack;
+    @Getter
+    private final boolean transferMaxCharge;
+    @Getter
+    private final boolean overrideCharge;
+
+
+    public ShapedEnergyTransferRecipe(ResourceLocation id, String group, int width, int height, ItemStack chargeStack, boolean overrideCharge, boolean transferMaxCharge, NonNullList<Ingredient> recipeItems, ItemStack result) {
+        super(id, group, width, height, recipeItems, result);
+        this.chargeStack = chargeStack;
+        this.transferMaxCharge = transferMaxCharge;
+        this.overrideCharge = overrideCharge;
+    }
+
+    @Override
+    public boolean matches(CraftingContainer inv, Level level) {
+        this.craftingContainer = inv;
+        return super.matches(inv, level);
+    }
+
+    @Override
+    public ItemStack getResultItem() {
+        long maxCharge = 0L;
+        long charge = 0L;
+        ItemStack resultStack = super.getResultItem();
+        for (int i = 0; i < craftingContainer.getContainerSize(); i++) {
+            if (craftingContainer.getItem(i).sameItemStackIgnoreDurability(chargeStack)) {
```

Comment:
This part, you're testing for stack equality here

It can very well be multiple different things you might want here. So use an ingredient